drivers/usb/udc/udc_bflb_bl61x.c

```diff
@@ -0,0 +1,1473 @@
+			buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, 0U);
+			err = udc_ctrl_submit_status(dev, buf);
+			udc_ctrl_update_stage(dev, buf);
+			net_buf_unref(buf);
+			return err;
+		}
+	} else {
+		LOG_ERR("Completed VDMA transfer of Unknown Stage");
+	}
+
+	return 0;
+}
+
+static uint32_t udc_bflb_bl61x_ep_remain(const struct device *dev, const uint8_t fifo_idx)
+{
+	const struct udc_bflb_bl61x_config *cfg = dev->config;
+	uint32_t tmp;
+
+	tmp = (sys_read32(cfg->base + USB_VDMA_F0PS1_OFFSET
+		+ (fifo_idx - 1) * USB_BL61X_FX_X_OFFSET)
+		& USB_VDMA_LEN_CXF_MASK);
+
+	return (tmp >> USB_VDMA_LEN_CXF_SHIFT);
+}
+
+static void udc_bflb_bl61x_ep_dout_start(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_bflb_bl61x_data *priv = udc_get_private(dev);
+	struct net_buf *buf;
+	uint8_t ep_idx = USB_EP_GET_IDX(ep_cfg->addr);
+
+	LOG_DBG("dout start ep 0x%02x", ep_cfg->addr);
+
+	if (priv->ep_is_in[ep_idx]) {
+		LOG_ERR("Unexpected ep 0x%02x dout token", ep_cfg->addr);
+	}
+
+	buf = udc_buf_peek(ep_cfg);
+	if (buf == NULL) {
+		LOG_ERR("No buffer for ep 0x%02x", ep_cfg->addr);
+		udc_submit_event(dev, UDC_EVT_ERROR, -ENOBUFS);
+	} else {
+		priv->ep_is_in[ep_idx] = false;
+		udc_bflb_bl61x_vdma_startread(dev,
+			udc_bflb_bl61x_ep_get_fifo(ep_cfg), buf->data, buf->size);
+	}
+}
+
+static void udc_bflb_bl61x_ep_din_start(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_bflb_bl61x_data *priv = udc_get_private(dev);
+	struct net_buf *buf;
+	uint8_t ep_idx = USB_EP_GET_IDX(ep_cfg->addr);
+
+	LOG_DBG("din start ep 0x%02x", ep_cfg->addr);
+
+	if (!priv->ep_is_in[ep_idx]) {
+		LOG_ERR("Unexpected ep 0x%02x din token", ep_cfg->addr);
+	}
+
+	buf = udc_buf_peek(ep_cfg);
+	if (buf == NULL) {
+		LOG_ERR("No buffer for ep 0x%02x", ep_cfg->addr);
+		udc_submit_event(dev, UDC_EVT_ERROR, -ENOBUFS);
+	}
+
+	LOG_DBG("start DMA for buf %p, data %p, len %i", (void *)buf, (void *)buf->data, buf->len);
+	priv->ep_is_in[ep_idx] = true;
+	udc_bflb_bl61x_vdma_startwrite(dev,
+		udc_bflb_bl61x_ep_get_fifo(ep_cfg), buf->data, buf->len);
+}
+
+static int udc_bflb_bl61x_ep_evt_end(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct net_buf *buf;
+	uint32_t remain = 0;
+
+	buf = udc_buf_get(ep_cfg);
+	LOG_DBG("Event end for 0x%02x got buf %lx", ep_cfg->addr, (uintptr_t)buf);
+	if (buf == NULL) {
+		return -ENODATA;
+	}
+	if (USB_EP_DIR_IS_OUT(ep_cfg->addr)) {
+		remain = udc_bflb_bl61x_ep_remain(dev, udc_bflb_bl61x_ep_get_fifo(ep_cfg));
+		net_buf_add(buf, ep_cfg->mps - remain);
+	} else {
+		net_buf_pull(buf, buf->len);
+	}
+
+	return udc_submit_ep_event(dev, buf, 0);
+}
+
+static void udc_bflb_bl61x_work_handler(struct k_work *item)
+{
+	struct udc_bflb_bl61x_ev *ev;
+	struct udc_ep_config *ep_cfg;
+	int err = 0;
+
+	ev = CONTAINER_OF(item, struct udc_bflb_bl61x_ev, work);
+
+	LOG_DBG("dev %p, ep 0x%02x, event %u",
+		ev->dev, ev->ep_addr, ev->event);
+
+	ep_cfg = udc_get_ep_cfg(ev->dev, ev->ep_addr);
```

Comment:
If applying the "const everywhere", then this could be refactored like this:

```suggestion
	struct udc_bflb_bl61x_ev *const ev = CONTAINER_OF(item, struct udc_bflb_bl61x_ev, work);
	struct udc_ep_config *const ep_cfg = udc_get_ep_cfg(ev->dev, ev->ep_addr);
	int err = 0;

	LOG_DBG("dev %p, ep 0x%02x, event %u",
		ev->dev, ev->ep_addr, ev->event);
```

Reply:
i think i got most of it...

Reply:
Thanks! If really wanting to convert *all* I've annotated them, but I don't see current drivers being applying this 100% either.